### ISSUE TITLE
Assertion failed: (castIsValid(op, S, Ty) && "Invalid cast!")

### ISSUE THREAD

**Issue by brooksdavis** [CTSRD-CHERI/clang#19]
https://gist.github.com/brooksdavis/ae374cfde7b32c969a01

```
/home/bed22/git/llvm-cheri/Build/bin/clang -integrated-as --target=cheri-unknown-freebsd  -msoft-float --sysroot=/home/bed22/obj/mips.mips64/home/bed22/git/cheribsd/tmp -mabi=sandbox  -O -pipe -G0   -I/home/bed22/git/cheribsd/lib/libexpat -DHAVE_EXPAT_CONFIG_H  -std=gnu99 -Wsystem-headers -Werror -Wall -Wno-format-y2k -Wno-uninitialized -Wno-pointer-sign -Wno-empty-body -Wno-string-plus-int -Wno-unused-const-variable -Wno-tautological-compare -Wno-unused-value -Wno-parentheses-equality -Wno-unused-function -Wno-enum-conversion -Wno-unused-local-typedef -Wno-switch -Wno-switch-enum -Wno-knr-promoted-parameter -Qunused-arguments -c /home/bed22/git/cheribsd/lib/libexpat/../../contrib/expat/lib/xmlparse.c -o xmlparse.o
Assertion failed: (castIsValid(op, S, Ty) && "Invalid cast!"), function Create, file ../lib/IR/Instructions.cpp, line 2316.
clang-3.6: error: unable to execute command: Segmentation fault (core dumped)
clang-3.6: error: clang frontend command failed due to signal (use -v to see invocation)
clang version 3.7.0 (https://github.com/llvm-mirror/clang.git 130f82c3373e3c728766025bd86d53cac22ba210) (https://github.com/llvm-mirror/llvm.git 1dc272f161841baef9f62af3e810a5cc68c9e1ae)
Target: cheri-unknown-freebsd
Thread model: posix
clang-3.6: note: diagnostic msg: PLEASE submit a bug report to http://llvm.org/bugs/ and include the crash backtrace, preprocessed source, and associated run script.
clang-3.6: note: diagnostic msg: 
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang-3.6: note: diagnostic msg: /tmp/xmlparse-6b97a4.c
clang-3.6: note: diagnostic msg: /tmp/xmlparse-6b97a4.sh
clang-3.6: note: diagnostic msg: 

********************
*** Error code 254
```

**Comment by brooksdavis** [comment]
Another example in libutil https://gist.github.com/brooksdavis/1755109e5c6ac8e11fdd

**Comment by brooksdavis** [comment]
Three more examples of this assert:

libutil: https://gist.github.com/brooksdavis/4203b52417dbc750ac4d
librt: https://gist.github.com/brooksdavis/0dd86b31c6f075c1f223
libexpat: https://gist.github.com/brooksdavis/2eab49e712459492eff9

**Comment by davidchisnall** [comment]
These are almost certainly different bugs.  It would help if you could retry the build command (the output from the Program Arguments: bit of the trace) in the debugger and file different bugs for each of the places where it's caused (step out of IRBuilder to whatever is invoking it).  For the most part, they're things that will be calling CreateBitcast and will want a fix similar to 3436e19.
